### ISSUE TITLE
Image preview in media_folder browser broken whith LFS enabled on repo

### ISSUE THREAD

**Issue by andytough** [netlify/netlify-cms#1206]
Thank you for the Netlify CMS. I am experimenting with it at the moment and I am frequently delighted as I explore the way it works! 

Previously when using the Netlify CMS for image selection the media browser would display thumbnails of uploaded images to select from.

I have just enabled LFS (large file support) for my repo and since then the media browser displays empty thumbnails.

Selected images do work correctly and are deployed successfully, it is only the media browser thumbnail that is broken.

If I copy the path of one of the broken thumbnails it is the path to the 'text pointer' for the image file, which makes sense.

Is there a configuration change I need to make to make this work as I would like it to, or, if not, would it be possible to make the 'text pointer' resolve to the actual image when using the media picker?

Example image: <https://raw.githubusercontent.com/andytough/dev-andytough-website/master/static/images/andy-tough_bwv3-ncms.jpg>

Netlify CMS config file: <https://github.com/andytough/dev-andytough-website/blob/master/static/admin/config.yml>
 
I am using Hugo, and the version, along with LFS enable is in the Netlify TOML file:
Netlify TOML: <https://github.com/andytough/dev-andytough-website/blob/master/netlify.toml>

**Comment by erquhart** [comment]
Could you give an example of the url we _should_ be using in this case?

**Comment by andytough** [comment]
Thank you for your response. I have modified my post, removing some of the links, and editing the content, in the hope that it will make more sense.

**Comment by tech4him1** [comment]
@andytough Is there an API you know of that we can use to fetch the actual LFS data? I've been looking, but haven't found one yet.

**Comment by andytough** [comment]
@tech4him1 Sorry, I don't know of any. I am quite new to this!

**Comment by erquhart** [comment]
Yeah, took a look myself and there doesn't seem to be an API for LFS. Might be worth asking GitHub on Twitter or in a forum.

**Comment by andytough** [comment]
@erquhart Thank you. I will see what I can find. Shall I change this issue from _Bug_ to _Feature Request_, as I think from your responses that what I am experiencing is expected behaviour?

**Comment by oncletom** [comment]
I +1 on this one as git-lfs can also be used to manage large assets, like PDFs, videos and such. I suspect these elements could have remote metadata if they are indexed in a documentstore somewhere (like Algolia, ElasticSearch, etc.)

**Comment by lookea** [comment]
Any news on github side on this ?

**Comment by erquhart** [comment]
We now support git lfs for [Netlify Large Media](https://www.netlifycms.org/docs/netlify-large-media/), not certain what that does for lfs in general. We should at least be close, all of the code is there for working with pointer files.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by lookea** [comment]
thanks @erquhart  yes Netlify own storage works, but when trying to use github lfs, all commits when uploading images are not lfs references but the image in self (bytes), this should be able to be handle I believe easily within the cms code, and I will really appreciate if that is done (I do not see a way to workaround this in the configuration) I know that other stuff like having the previews working requires implementation and github http server details.

**Comment by lookea** [comment]
> We now support git lfs for [Netlify Large Media](https://www.netlifycms.org/docs/netlify-large-media/), not certain what that does for lfs in general. We should at least be close, all of the code is there for working with pointer files.
> 
> Sent with [GitHawk](http://githawk.com)

Hi @erquhart I can confirm netlify own lfs solution works, but there is a problem, only pages that are merge into master can render the image preview, when using editorial workflow when creating a draft all images are shown in the preview but after going out and back into it the images are then out of the cache and no longer rendered, they only show if you go from draft to publish, merge into master, I think this is some kind of issue with netlify-cms code because images are indeed available on the lfs

**Comment by erezrokah** [comment]
Linking https://github.com/netlify/netlify-cms/pull/3106 as reference

**Comment by lookea** [comment]
What is the https://app_name.netlify.app/.netlify/large-mediaapi call done ? that seems to work together with git-gateway auth, it has a /verify resource,  and why it is different from the lfs service url that is something like https://XXXXXXXXXX.netlify.com/.netlify/large-media very similar structure. I don't see that webservice anywere on the github repository of netlify, and it doesn't follow the git lfs spec, so is something different

**Comment by erezrokah** [comment]
Hi @lookea, those calls are equivalent expect using a different domain name for the site. The API stays the same.
The way Git LFS works is by uploading files that match a certain patten to a 3rd party service (Netlify Large Media in this case) instead of saving them directly in the repo.
What's saved in the repo is a pointer file containing information that can be used to extract the file from the 3rd party service.
Netlify CMS detects if you're using Netlify Large Media and automatically uploads matching files to that service using dedicated API methods like `/verify`.
To support `Git-LFS` for GitHub, the GitHub API would need to provide similar API methods to facilitate that functionality.

**Comment by srchulo** [comment]
@erezrokah so even though GitHub supports LFS, it's missing some API call to upload media as LFS through Netlify CMS? Can't Netlify CMS add media files by committing using LFS?

**Comment by erezrokah** [comment]
> Can't Netlify CMS add media files by committing using LFS?

That would require the CMS to have a JavaScript implementation of the `git-lfs` protocol and also has some challenges on using the OAuth token for such operations.

Of course I'm open to someone from the community looking into this.

**Comment by lookea** [comment]
I understand how it works so it is dependent on the oauth handling of git-gateway in order to support pushing the lfs files, anyway is there any reason why the code for the large-media endpoint to not be public available ? I think having that I can try to adapt it to git lfs, and work on the required cms js stuff side given to me it is only supporting netlify lfs only

**Comment by lookea** [comment]
> @erezrokah so even though GitHub supports LFS, it's missing some API call to upload media as LFS through Netlify CMS? Can't Netlify CMS add media files by committing using LFS?

Is sad but the only direct way of communication happens with git backend implementation, and even there given how github works you require a middle man to get the oauth token, that is ok  given it can easily be implemented, netlify provide it out of the box of course, but LFS is a little more complex, given you have more auth stuff to handle, netlify does this by using the git-gateway a more complete and complex solution that handles username/password authentication as a bridge to get github access. But yes the magic of having github as a database was to eliminate the need of a backend for a lot of small projects.

**Comment by erezrokah** [comment]
I think this discussion is getting a bit out of focus.

You don't need `git-gateway` to have LFS.

We recently had a community contributed PR to support LFS for Bitbucket - that was possible since Bitbucket exposes its LFS API.

Since GitHub doesn't expose an LFS API adding such an implementation is not straightforward when used directly.

A possible solution at the moment (not involving Netlify open sourcing Netlify Large Media service code or switching to Bitbucket) will be to write your own backend and hook into media related operations to write media files to any destination other than GitHub.

I agree that is not ideal, but I don't see any easy fix to this issue (unless API access is available).

Regardless, I don't think we want the core `github` backend to be tied to any LFS service that is not provided by `github`.

**Comment by naavis** [comment]
I guess this is not going to happen any time soon, given no one has commented in over a year?

**Comment by erezrokah** [comment]
Hi @naavis, unless GitHub added LFS support to their API I don't see a way forward.
Happy to accept a contribution for it  if someone figures out a solution.

**Comment by JohannesFerner** [comment]
+ Please add to the docs of github backend, that LFS is not supported for now, as it's a pain to revert those changes in a live-repository which is actively maintained by content-editors.

**Comment by erezrokah** [comment]
Hi @JohannesFerner, we'd be happy to receive a docs contribution for it.
There's an `Edit this page` button one can use to submit a change.

**Comment by JohannesFerner** [comment]
> Hi @JohannesFerner, we'd be happy to receive a docs contribution for it. There's an `Edit this page` button one can use to submit a change.

thanks for the hing - tried that now- doesnt work btw. 

![Content_Manager](https://user-images.githubusercontent.com/1781395/152148973-b2cd78e3-b4c7-4a04-94b3-9d0c2e380800.jpg)

MD to add:
```
## Git Large File Storage (LFS)

Please note that the Github-Backend configuration is **not yet** supporting [git-lfs](https://git-lfs.github.com/), see [\#1206](https://github.com/netlify/netlify-cms/issues/1206) for details and status.
```

**Comment by erezrokah** [comment]
Thanks @JohannesFerner, not sure why it's not working. I created https://github.com/netlify/netlify-cms/pull/6175 and will merge soon

**Comment by lukebouch** [comment]
To be clear, does this work with Netlify large media?

**Comment by erezrokah** [comment]
> To be clear, does this work with Netlify large media?

Yes with the `git-gateway` backend. This issue is a feature request to add the same functionally for the `github` backend.

**Comment by 3x071c** [comment]
> > To be clear, does this work with Netlify large media?
> 
> Yes with the `git-gateway` backend. This issue is a feature request to add the same functionally for the `github` backend.

Netlify Large Media won't work with the `GitHub` backend?

**Comment by JohnJMills** [comment]
Wondering about the status of this issue. Are there any plans to pick this up?

**Comment by jezcope** [comment]
Netlify Large Files has been [deprecated since 2023](https://answers.netlify.com/t/large-media-feature-deprecated-but-not-removed/100804) and cannot be added to new sites. It would be good to have an alternative!

**Comment by swt2c** [comment]
I started to take a look into this.  GitHub does have an API for accessing LFS it seems - the Git LFS API.  However, it runs on lfs.github.com, which unfortunately does not seem to allow CORS requests like api.github.com does.  So I think that's the main blocker.
